onnxruntime/core/providers/cuda/reduction/reduction_functions.cc

```diff
@@ -13,45 +15,86 @@
 namespace onnxruntime {
 namespace cuda {
 
-ApplicableMatrixReduction get_applicable_matrix_reduction(
-    const cudnnReduceTensorOp_t cudnn_reduce_op,
-    const std::vector<int64_t>& dims, const std::vector<int64_t>& original_axes,
-    int& m_out, int& n_out) {
-  if (cudnn_reduce_op != CUDNN_REDUCE_TENSOR_ADD) {
-    return ApplicableMatrixReduction::None;
+namespace {
+// gets min and max of single contiguous range of axes if available
+optional<std::pair<int64_t, int64_t>> GetMinMaxContiguousAxes(
+    int64_t rank,
+    const std::vector<int64_t>& dims,
+    const std::vector<int64_t>& original_axes) {
+  // empty axes means reduce all dimensions
+  if (original_axes.empty()) {
+    return std::make_pair(int64_t{0}, rank - 1);
   }
 
-  const auto rank = gsl::narrow<int64_t>(dims.size());
+  // normalize axis values and sort
+  const std::vector<int64_t> axes = [&original_axes, rank]() {
+    std::vector<int64_t> result(original_axes);
+    std::for_each(
+        result.begin(), result.end(),
+        [rank](int64_t& axis) { axis = HandleNegativeAxis(axis, rank); });
+    std::sort(result.begin(), result.end());
+    return result;
+  }();
+
+  assert(!axes.empty());
```

Comment:
Aldo do `assert(dims.size() == rank)` for safety.

Reply:
added above